### PR TITLE
Add Cognito

### PR DESCRIPTION
**Pull request by JamesSiebert**
Adding support for AWS Cognito OAuth2 connections.
Includes support for logging out of Cognito.

### CODE REVIEW

File: src/Cognito/Provider.php [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    // Logout the user from cognito (expire SSO tokens), then redirect to url.
+    public function logoutCognitoUser()
+    {
+        $authHost = config('services.cognito.host');
+        $clientId = config('services.cognito.client_id');
+        $logoutUri = config('services.cognito.logout_uri');
+
+        return "$authHost/logout?client_id=$clientId&logout_uri=$logoutUri";
```
Comment (atymic) [inline src/Cognito/Provider.php:157]: could we use sprintf here please

File: src/Cognito/Provider.php [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    protected $scopeSeparator = ' ';
+
+    /**
+     * {@inheritdoc}
+     */
+    public static function additionalConfigKeys()
```
Comment (atymic) [inline src/Cognito/Provider.php:42]: Chuck this at the end of the class please (just makes it a little easier to read)

File: src/Cognito/Provider.php [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    protected function getTokenFields($code)
+    {
+        return array_merge(parent::getTokenFields($code), [
+            'grant_type' => 'authorization_code',
+        ]);
+    }
+
+    protected function getCognitoUrl($type)
+    {
+        return rtrim($this->getConfig('host'), '/').'/'.ltrim(($this->getConfig($type, Arr::get([
```
Comment (atymic) [inline src/Cognito/Provider.php:138]: This statement is a little confusing, could we split it out into a couple of statements with a sprintf to concat together?

**Review by atymic — COMMENTED** [review]
Awesome work, just a few minor bits!

File: src/Cognito/Provider.php [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+            'host',
+            'authorize_uri',
+            'token_uri',
+            'userinfo_uri',
+            'userinfo_key',
+            'user_id',
+            'user_nickname',
+            'user_name',
+            'user_email',
+            'user_avatar',
```
Comment (lucasmichot) [inline src/Cognito/Provider.php:54]: Are you sure those other fields, are actually part of the config ?

File: src/Cognito/Provider.php [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+     * Get the raw user for the given access token.
+     *
+     * @param string $token
+     *
+     * @return array
+     */
+    protected function getUserByToken($token)
+    {
+        $response = $this->getHttpClient()->get($this->getCognitoUrl('userinfo_uri'), [
+            'headers' => [
```
Comment (lucasmichot) [inline src/Cognito/Provider.php:90]: Can you use `RequestOption::HEADERS` here instead?

File: src/Cognito/Provider.php [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+                'Authorization' => 'Bearer '.$token,
+            ],
+        ]);
+
+        return (array) json_decode($response->getBody(), true);
```
Comment (lucasmichot) [inline src/Cognito/Provider.php:95]:
```suggestion
        return json_decode((string) $response->getBody(), true);
```

File: src/Cognito/Provider.php [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+
+    protected function getUserData($user, $key)
+    {
+        return Arr::get($user, $this->getConfig('user_'.$key, $key));
+    }
+
+    // Logout the user from cognito (expire SSO tokens), then redirect to url.
+    public function logoutCognitoUser()
+    {
+        $authHost = config('services.cognito.host');
```
Comment (lucasmichot) [inline src/Cognito/Provider.php:153]:
```suggestion
        $authHost = $this->getConfig('host');
```

File: src/Cognito/Provider.php [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+        $clientId = config('services.cognito.client_id');
```
Comment (lucasmichot) [inline src/Cognito/Provider.php:154]:
```suggestion
        $clientId = $this->getConfig('client_id');
```

File: src/Cognito/Provider.php [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+        $logoutUri = config('services.cognito.logout_uri');
```
Comment (lucasmichot) [inline src/Cognito/Provider.php:155]:
```suggestion
        $logoutUri = $this->getConfig('logout_uri');
```

File: src/Cognito/Provider.php [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+
+        return "$authHost/logout?client_id=$clientId&logout_uri=$logoutUri";
+    }
+
+    // Logout the user from cognito (expire SSO tokens), then redirect the user to he sign in page.
+    public function switchCognitoUser()
```
Comment (lucasmichot) [inline src/Cognito/Provider.php:161]: missing `@return` docblock

File: src/Cognito/composer.json [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+            "email": "js@media360.com.au"
+        }
+    ],
+    "require": {
+        "php": "^7.2 || ^8.0",
+        "ext-json": "*",
+        "socialiteproviders/manager": "~4.0"
+    },
+    "require-dev": {
+        "mockery/mockery": "~0.9.4",
```
Comment (lucasmichot) [inline src/Cognito/composer.json:24]: unused.

Check the `composer.json` files for the other providers.

**Review by lucasmichot — CHANGES_REQUESTED** [review]
Great addition James !

Still some changes required

File: src/Cognito/Provider.php (hunk 1 above)
Comment (JamesSiebert) [reply src/Cognito/Provider.php]: done

File: src/Cognito/Provider.php (hunk 3 above)
Comment (JamesSiebert) [reply src/Cognito/Provider.php]: done.

Comment (JamesSiebert) [reply src/Cognito/Provider.php]: Thanks for the feedback :)

File: src/Cognito/Provider.php (hunk 4 above)
Comment (JamesSiebert) [reply src/Cognito/Provider.php]: Fixed - this was referenced from another provider, I understand it better now and have updated it.

File: src/Cognito/Provider.php (hunk 5 above)
Comment (JamesSiebert) [reply src/Cognito/Provider.php]: Done.

File: src/Cognito/Provider.php (hunk 6 above)
Comment (JamesSiebert) [reply src/Cognito/Provider.php]: Done.

File: src/Cognito/Provider.php (hunk 7 above)
Comment (JamesSiebert) [reply src/Cognito/Provider.php]: Done.

File: src/Cognito/Provider.php (hunk 8 above)
Comment (JamesSiebert) [reply src/Cognito/Provider.php]: Done.

File: src/Cognito/Provider.php (hunk 9 above)
Comment (JamesSiebert) [reply src/Cognito/Provider.php]: Done.

File: src/Cognito/Provider.php (hunk 10 above)
Comment (JamesSiebert) [reply src/Cognito/Provider.php]: I think this is now done, I referenced another provider.

File: src/Cognito/composer.json (hunk 11 above)
Comment (JamesSiebert) [reply src/Cognito/composer.json]: Removed, thanks for the feedback :)

**Comment by JamesSiebert** [comment]
> Great addition James !
> 
> Still some changes required

I've implemented the requested changes and It feels a lot cleaner now.
Thanks for the suggestions, let me know if you have any more suggestions :)

**Comment by atymic** [comment]
@lucasmichot I see you PR'd, are you happy with the current state?

**Comment by lucasmichot** [comment]
> @lucasmichot I see you PR'd, are you happy with the current state?

Let's wait for for @JamesSiebert to merge https://github.com/JamesSiebert/Providers/pull/1 
Once done, this PR will be good to go

**Review by atymic — APPROVED** [review]

**Pull request merged**